sphinx/themes/basic/static/searchtools.js

```diff
@@ -510,9 +511,27 @@ const Search = {
       if (word.length > 2) {
         const escapedWord = _escapeRegExp(word);
         if (!terms.hasOwnProperty(word)) {
-          Object.keys(terms).forEach((term) => {
-            if (term.match(escapedWord))
-              arr.push({ files: terms[term], score: Scorer.partialTerm });
+          const termOffsets = Object.keys(terms);
+          const ngramTerms = function (ngram) {
+            let [node, path] = [termsNgrams, ""];
+            for (const step of ngram) {
+              if ((path += step) in node) [node, path] = [node[path], ""];
+            }
+            if (!node) return [];
+            if (node.length === undefined) node = [node];
+            return node.map(offset => termOffsets[offset]);
+          };
+
+          const candidateTerms = new Set(ngramTerms(word.substring(0, 3)));
+          for (let start = 1; candidateTerms.size && start + 3 <= word.length; start++) {
```

Comment:
```suggestion
          for (let start = 1; candidateTerms.size && start + 3 <= word.length; start++) {
```

Is it valid to skip-ahead by two characters (or in the general case: `len(ngram) - 1`) here?

That would mean that for a query word `foobar`, we find terms that contain all of:

- `foo`, `oba`

Instead of:

- `foo`, `oob`, `oba`, `bar`

So, to answer my own question: _not quite_.  Yes, we can skip ahead `len(ngram) - 1` for most matches, but we do also need to make sure that we check for an end-of-word match.  What we want in the example case is:

- `foo`, `oba`, `bar`

Comparisons, loops and set operations can add up in terms of computational cost, so despite seeming like an overoptimization (it does add some complexity), this may be worthwhile to consider.

Reply:
A symmetric way to do this could be to start checking for indexed ngrams that match at both the start and end of the query word - and to advance both of those checks inwards towards the middle of the word by `len(ngram) - 1` on each iteration.

Reply:
> A symmetric way to do this could be to start checking for indexed ngrams that match at both the start and end of the query word - and to advance both of those checks inwards towards the middle of the word by `len(ngram) - 1` on each iteration.

(stopping when the two substring ranges intersect -- and with an optimization that only one lookup is required when the ranges are identical)

Reply:
There could be some improvements here, but from experimenting with them locally: I haven't been able to implement them in a way that I consider readable and maintainable, at least not with my level of JavaScript proficiency.  So I'll hold off on this.